Doc: Generic Http Client (#452)
SAP/cloud-sdk: docs/js/features/connectivity/on-premise.md

---
title: On-Premise Connectivity in the SAP Cloud SDK for JavaScript
sidebar_label: On-Premise Systems

> description: This article describes how the SDK request to On-Premise SAP S/4 HANA systems.

Reviewer: ```suggestion
description: How to connect to SAP S/4 HANA On-Premise systems using the SAP Cloud SDK
```

keywords:
- sap
- cloud
- sdk
- On-Premise
- connectivity
- JavaScript
- TypeScript
---

## Introduction ##

> This document has overlap to the [proxy handling](./proxy.md), [destination lookup](./destination.md) and [On-Premise connectivity from the BAS](../../guides/bas-external-system.md).
> The focus of this document is on the steps necessary to connect to an On-Premise system and how the SDK implements these steps.
> The crucial problem with On-Premise SAP S/4 HANA systems is, that they are not exposed to the internet.

Reviewer: These first lines sound more scientific than I would expect in an online SDK documentation. Maybe you can motivate it the other way around and leave out "the focus of this document". This should become clear from the description.

```suggestion
The crucial problem with SAP S/4HANA On-Premise systems is, that they are not exposed to the internet.
The SAP Cloud SDK has some useful features that allow you to connect to On-Premise systems.
```

(This sounds a bit dry, but I guess you get what I mean).

Author: i adjusted the sentences and added a picture

> Typically, they are located within the companies internal network.

Reviewer: ```suggestion
Typically, they are located within a company's internal network.
```

> Since this isolation would be in conjunction to modern requirements of data availability, SAP has introduced the [Cloud Connector](https://help.sap.com/viewer/cca91383641e40ffbe03bdc78f00f681/Cloud/en-US/e6c7616abb5710148cfcf3e75d96d596.html?q=cloud%20connector).

Reviewer: What do you mean by this?
"would be in conjunction to" => "contradicts the"?

Author: yes

> The Cloud Connector works like a reverse proxy opening a connection between some white-listed service endpoints of the On-Premise system, and one Cloud Foundry account.

Reviewer: ```suggestion
The Cloud Connector works like a reverse proxy opening a connection between some white-listed service endpoints of the On-Premise system, and the connected Cloud Foundry account.
```

Only applications running in the connected Cloud Foundry account can access the services. 

> Applications which are supposed to access On-Premise systems need a binding to the [connectivity service.](https://www.cloudfoundry.org/the-foundry/sap-cloud-platform-service-connectivity/)

Reviewer: ```suggestion
Applications which are supposed to access On-Premise systems need a binding to the [connectivity service](https://www.cloudfoundry.org/the-foundry/sap-cloud-platform-service-connectivity/).
```

> ## Implementation Details ##

Reviewer: This title is not very inspiring. But to be fair I don't have anything better in mind.

> As a first step, the SDK looks up the destination as described in [this document](./destination.md).

Reviewer: ```suggestion
As a first step, the SDK [looks up the destination](./destination.md).
```

If the `Proxy Type` of the destination is `OnPremise` the SDK will try to establish a connection via the Cloud Connector.

As mentioned above, the application needs a service binding to the connectivity service.

> A service binding relates to client credentials in the  `VCAP` environment variables. 

Reviewer: "relates to" is quite unspecific

Author: made it more concrete

> A service binding relates to client credentials in the  `VCAP` environment variables. 

Reviewer: ```suggestion
A service binding relates to client credentials in the  `VCAP` environment variable. 
```

The SDK uses the client credentials and if given the tenant id of the JWT to obtain a [client credentials grant.](https://help.sap.com/viewer/8d8be6a74e4e49589a546c02ee193741/latest/en-US/f1eff1dd7907469491989b3a36e6a7c6.html)
If multiple connectivity services binding exist, the SDK uses the first entry per default.

The SDK takes the client grant to call the connectivity service and receives the host, password and authorization of the connectivity proxy.
Then the SDK creates an HTTP agent considering this proxy and the necessary `Proxy-Authorization` headers.
The proxy is the entry point to the Cloud Connector instance connected the account.
The proxy is only reachable from the Cloud Foundry space, so you cannot use that flow for applications running outside of Cloud Foundry. 

> These are the main steps the SDK does to achieve connectivity to SAP S/4 HANA On-Premise systems via the Cloud Connector.

Reviewer: I think you can strike this sentence. Maybe you can use this sentence to find a more speaking title for this section?
```suggestion
These are the main steps the SDK does to achieve connectivity to SAP S/4 HANA On-Premise systems via the Cloud Connector.
```

> We would like to mention to minor aspects of the SDK:

Reviewer: ```suggestion
We would like to mention minor aspects of the SDK:
```

> We would like to mention to minor aspects of the SDK:

Reviewer: For the part below I would make an extra section or subsection - or maybe even two subsections, because then it will be mentioned in the TOC on the side. If you find a nice title this might be helpful.

Author: yes good idea

---

> title: On-Premise Connectivity in the SAP Cloud SDK for JavaScript / TypeScript

Reviewer 2: use the official name and also the longer name occupies 3 lines for the title
```suggestion
title: On-Premise Connectivity in the SAP Cloud SDK for JavaScript
```

hide_title: false
hide_table_of_contents: false

> sidebar_label: Proxies

Author: oh yes

> description: This article describes how the SDK request to On-Premise sAP S/4 HANA systems.

Reviewer 2: ```suggestion
description: This article describes how the SDK request to On-Premise SAP S/4 HANA systems.
```

The focus of this document is on the steps necessary to connect to an On-Premise system and how the SDK implements these steps.
The crucial problem with On-Premise SAP S/4 HANA systems is, that they are not exposed to the internet.
Typically, they are located within the companies internal network.

Since this isolation would be in conjunction to modern requirements of data availability, SAP has introduced the [Cloud Connector](https://help.sap.com/viewer/cca91383641e40ffbe03bdc78f00f681/Cloud/en-US/e6c7616abb5710148cfcf3e75d96d596.html?q=cloud%20connector).
The Cloud Connector works like a reverse proxy opening a connection between some white-listed service endpoints of the On-Premise system, and one Cloud Foundry account.
Only applications running in the connected Cloud Foundry account can access the services. 
Applications which are supposed to access On-Premise systems need a binding to the [connectivity service.](https://www.cloudfoundry.org/the-foundry/sap-cloud-platform-service-connectivity/)

## Implementation Details ##

> As a first step the SDK looks up the destination as described in [this document](./destination.md).

Reviewer 2: ```suggestion
As a first step, the SDK looks up the destination as described in [this document](./destination.md).
```

If the `Proxy Type` of the destination is `OnPremise` the SDK will try to establish a connection via the Cloud Connector.

> As mentioned above the application needs a service binding to the connectivity service.

Reviewer 2: ```suggestion
As mentioned above, the application needs a service binding to the connectivity service.
```

If multiple connectivity services binding exist, the SDK uses the first entry per default.

The SDK takes the client grant to call the connectivity service and receives the host, password and authorization of the connectivity proxy.
Then the SDK creates an HTTP agent considering this proxy and the necessary `Proxy-Authorization` headers.
The proxy is the entry point to the Cloud Connector instance connected the account.
The proxy is only reachable from the Cloud Foundry space, so you cannot use that flow for applications running outside of Cloud Foundry. 

These are the main steps the SDK does to achieve connectivity to SAP S/4 HANA On-Premise systems via the Cloud Connector.
We would like to mention to minor aspects of the SDK:
1. In case the `Authentication` property on the destination is `PrincipalPropagation` the SDK adds the `SAP-Connectivity-Authentication` header to the request.
The header contains user information included in the `JWT` and enables user propagation from the cloud application to the SAP S/4 HANA system.
2. In case the destination contains a  `LocationId` property, the SDK adds the `SAP-Connectivity-SCC-Location_ID` header to the request.

> If you have multiple Cloud Connectors connected to your account the `LocationId` defines which one of these to use for the request.

Reviewer 2: ```suggestion
If you have multiple Cloud Connectors connected to your account, the `LocationId` defines which one of these to use for the request.
```
